### ISSUE TITLE
VLAN tagged frames are received without tag from cloud ethernet interface on Linux

### ISSUE THREAD

**Issue by ehlers** [GNS3/gns3-server#849]
GNS3 version 2.0.0dev7 on Debian Stretch (64-bit) with Python 3.5.2+ Qt 5.7.1 and PyQt 5.7.

Simple Topology:
![cloud_vlan](https://cloud.githubusercontent.com/assets/896406/21522873/998a9b54-cd09-11e6-9b5b-a6cce021bc0b.png)

VLAN tagged frames are received without tag from the ethernet interfaces

Wireshark on the ethernet interface (outside GNS3) shows, that the received frames have a VLAN tag
![cloud_sniff](https://cloud.githubusercontent.com/assets/896406/21522908/d4e37626-cd09-11e6-8347-e473fd52fcaf.png)

But the ubridge.log shows, that it is send with VLAN tag, but received without VLAN tag
```
Received 64 bytes on bridge fb750a8c-5a4c-4c6f-8dcd-1ee69a257413-0 (source NIO)
0000: ff ff ff ff ff ff c2 01 17 d7 00 00 81 00 00 64 ...............d
0010: 08 06 00 01 08 00 06 04 00 01 c2 01 17 d7 00 00 ................
0020: 0a 64 01 02 00 00 00 00 00 00 0a 64 01 64 00 00 .d.........d.d..
0030: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................

Received 56 bytes on bridge fb750a8c-5a4c-4c6f-8dcd-1ee69a257413-0 (destination NIO)
0000: c2 01 17 d7 00 00 c8 2a 14 21 cf c8 08 06 00 01 ................
0010: 08 00 06 04 00 02 c8 2a 14 21 cf c8 0a 64 01 64 .............d.d
0020: c2 01 17 d7 00 00 0a 64 01 02 00 00 00 00 00 00 .......d........
0030: 00 00 00 00 00 00 00 00                         ........
```

This does not happen on OS X, there this scenario works fine.

I assume, that this is an ubridge issue on Linux only. I've read, that on Linux since 2.6.21 the VLAN information is stored in an AUXDATA structure. I tried to dig a little bit deeper, but I must confess, that it looks too complicated for me.

**Comment by grossmj** [comment]
I think this is because we use RAW sockets on Linux (NIO_Linux in 1.5.x) in
2.0 by default  instead of the libpcap (NIO_Ethernet).

I know RAW sockets drop VLAN tags. Maybe there is a way around this,
otherwise we may have to use pcap by default if installed.

**Comment by ehlers** [comment]
Here a patch for pcap-linux how they add back the VLANs: http://seclists.org/tcpdump/2008/q3/35

The actual code in https://github.com/the-tcpdump-group/libpcap/blob/master/pcap-linux.c has even a bit more code to handle this. So it's doable, but needs some effort.

**Comment by ehlers** [comment]
That doesn't work at all.

* In nio_linux_raw.h I have to change uint16_t to u_int16_t, otherwise it won't compile
* With a printf, I verified, that indeed recvmsg is called
* This "for (cmsg = CMSG_FIRSTHDR(&msg) ..." loop doesn't find the VLAN tag
* The memmove call is plain wrong, because you have to move "pkt + ETH_ALEN * 2" to "pkt + ETH_ALEN * 2 + VLAN_HEADER_LEN" with length "received - ETH_ALEN * 2". And the received packet length has to be incremented by VLAN_HEADER_LEN. But because the VLAN tag is not found, this doesn't make any harm

**Comment by grossmj** [comment]
I had little hope it would work... They don't make it easy for sure :(

**Comment by ehlers** [comment]
Created pull request https://github.com/GNS3/ubridge/pull/24

Tested on Debian Stretch (Kernel 4.8.7) and OS X 10.11.6.
On OS X it worked before, I verified that the changes don't break it.

Besides bug fixing, I added some changes from the current pcap library
* better check, if VLAN is available
* support for tp_id other then ETH_P_8021Q

I don't think, that the two memset to clear msg and msg_buf are really necessary (pcap doesn't do that), but they don't harm, so I left them in.

What's needed is to check, if this compiles and run on older linux versions, e.g. TP_STATUS_VLAN_TPID_VALID (for support of tp_id other then ETH_P_8021Q) was introduced in v3.14.
